keps/sig-auth/2718-20210511-client-exec-proxy/README.md

```diff
@@ -0,0 +1,834 @@
+because no one is listening, client-go will reattempt to exec the plugin.
+Client-go will support both plugins binaries which are themselves the proxy
+process, and additionally plugin binaries which complete, but are responsible
+for starting a proxy process and returning a URL to the proxy.
+
+The client will send its request to the proxy plugin process, and the proxy
+will forward the request to the API server, rather than returning the modified
+request to the client to be sent (see alternatives considered).  The downside
+of this approach is that the proxy must duplicate the some of the transport
+functionality of client packages, like respecting existing proxy configuration,
+but allows for more flexibility for proxy plugin developers.
+
+### Authentication to Proxy
+
+When communicating to the proxy over tcp/ip, the authentication method to the
+proxy is certificate based. The proxy passes back a client certificate and
+certificate authority that the client can use on connection.
+
+When communicating to the proxy over a unix domain socket, the authentication
+method to the proxy is file permission based.  The GID owner of the client
+process is passed to the proxy over stdin, which the proxy should give
+read/write permissions on the socket to.
+
+### Caching
+
+The connection from the client to the proxy can be cached in the client based
+on the parameters in the ExecProxyRequest: the proxyURL and cluster details.
+
+### Proxy Shutdown
+
+In the case where the client dies unexpectedly, due to a crash or SIGKILL, the
+proxy should also die.  This could be accomplished with 2 approaches: the proxy
+could quit after a certain amount of idle time, or the proxy could attempt to
+take a file lock that the parent holds and quit if it ever successfully takes
+the lock.  Both are reasonable approaches, so idle quit will be used due to its
+simplicity.  Proxies that are idle for more than 5 minutes will be required to
+quit.
+
+### API
+
+The ExecConfig will be extended with a `ProxyConfig` object which signals to
+the client to follow the protocol outlined above, instead of execing the binary
+to get credentials.
+
+The communication over stdin and stdout with the proxy will be done with using
+ExecProxyRequest and ExecProxyResponse objects.
+
+```go
+type ExecConfig struct {
+	// Command to execute.
+	Command string `json:"command"`
+	// Arguments to pass to the command when executing it.
+	// +optional
+	Args []string `json:"args"`
+	// Env defines additional environment variables to expose to the process. These
+	// are unioned with the host's environment, as well as variables client-go uses
+	// to pass argument to the plugin.
+	// +optional
+	Env []ExecEnvVar `json:"env"`
+
+	// Preferred input version of the ExecInfo. The returned ExecCredentials MUST use
+	// the same encoding version as the input.
+	APIVersion string `json:"apiVersion,omitempty"`
+
+	// ...
+
+	// +optional
+	ProxyConfig ExecProxyConfig `json:"proxyConfig"`
```

Comment:
1. why use an empty struct as bool?
2. instead of adding a boolean value in ExecConfig, we could add a new `ExecConfig` field in `type AuthInfo struct {` (per-user) or `type Cluster struct {` (per-cluster) and call it proxy.

the kubeconfig would be something like this
```
- name: nic@cluster1
  user:
    proxy:
      apiVersion: client.authentication.k8s.io/v1alpha1
      args:
      - --region=us-west-2
      command: /usr/local/bin/sigv4proxy
```

Reply:
1. It is a placeholder for now, I removed the contents of the struct but left it until I came back to this piece.
2. Good suggestion.  I think per-user makes sense and aligns with a per-client-process proxy.